From 7f45ad8064d6cb4c68ce0f5e1327cf1e97cb8af5 Mon Sep 17 00:00:00 2001
From: Vit Mojzis <vmojzis@redhat.com>
Date: Wed, 5 Apr 2017 11:51:56 +0200
Subject: [PATCH 2/2] framework: improve obtaining AVC object path

There can be more than one PATH record accompanying an AVC. Try to
choose the one corresponding to the AVC object (avoid PARENT type PATH
record if possible and use path matching name field of the AVC record if
it exists).

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1343641

Signed-off-by: Vit Mojzis <vmojzis@redhat.com>
---
 framework/src/setroubleshoot/audit_data.py | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/framework/src/setroubleshoot/audit_data.py b/framework/src/setroubleshoot/audit_data.py
index 1860075..a48422c 100644
--- a/framework/src/setroubleshoot/audit_data.py
+++ b/framework/src/setroubleshoot/audit_data.py
@@ -785,7 +785,7 @@ class AVC:
         '''
 
         path = None
-        name = None
+        name = self.avc_record.get_field('name')
 
         # First try to get the path from the AVC record, new kernel
         # versions put it there rather than in AVC_PATH
@@ -795,10 +795,20 @@ class AVC:
             path = path.strip('"')
         inodestr = self.avc_record.get_field("ino")
         if path is None:
-            avc_path_record = self.audit_event.get_record_of_type('PATH')
-            if avc_path_record:
+            # No path field in AVC record, try to get path from PATH records
+            avc_path_records = self.audit_event.get_records_of_type('PATH')
+            for avc_path_record in avc_path_records:
+                record_type = avc_path_record.get_field('objtype')
+                #Avoid PARENT records if possible
+                if path and record_type == "PARENT":
+                    continue
+
                 path = avc_path_record.get_field('name')
-            
+                # If there is more non-PARENT PATH records, use the one matching
+                # the name from AVC record... otherwise use the last one
+                if path and name and record_type != "PARENT" and path.endswith(name):
+                    break
+
         if path is None:
             # No path field, so try and use the name field instead
             name = self.avc_record.get_field('name')
-- 
2.12.2

